api/src/main/java/io/opentelemetry/common/experimental/KeyedAttributeDemo.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common.experimental;
+
+import io.opentelemetry.common.experimental.KeyedAttributes.BooleanKey;
+import io.opentelemetry.common.experimental.KeyedAttributes.StringArrayKey;
+import io.opentelemetry.common.experimental.KeyedAttributes.StringKey;
+import io.opentelemetry.common.experimental.ReadableKeyedAttributes.AttributeConsumer;
+import java.util.List;
+
+@SuppressWarnings("JavadocMethod")
+public class KeyedAttributeDemo {
+  private KeyedAttributeDemo() {}
+
+  // here are some semantic attributes examples:
+  static final StringKey HTTP_METHOD = KeyedAttributes.stringKey("http.method");
+  static final StringKey NET_HOST_NAME = KeyedAttributes.stringKey("net.host.name");
+  static final BooleanKey ERROR_HINT = KeyedAttributes.booleanKey("error.hint");
+  static final StringArrayKey HTTP_HEADERS = KeyedAttributes.stringArrayKey("http.headers");
+
+  public static void main(String[] args) {
+    // here's how you build a full set of attributes. You can imagine the Span/Builder would look
+    // similar
+    KeyedAttributes keyedAttributes =
+        KeyedAttributes.newBuilder()
+            .set(HTTP_HEADERS, "x-stuff", "x-api-key")
+            .set(HTTP_METHOD, "PUT")
+            .set(ERROR_HINT, false)
+            .set(NET_HOST_NAME, "localhost")
+            .build();
+
+    // iterate over them like this, using the readable interface.
+    keyedAttributes.forEach(
+        new AttributeConsumer() {
```

Comment:
I think this bothers me a bit. I'd suggest having a consumer only care about a single generic type and skip anything not of that type. If you need to iterate over everything you'd declare the root type.

Reply:
I agree. I was pondering a catch-all `consume` method for the unknown keys/values.

Reply:
I added the catch-all method, just for you. ;)

Reply:
@tylerbenson Just to see what an API might look like, I added support for setting custom attribute types, as well. I'm not sure how an SDK might use these attributes, or how you might implement your "compound attribute" idea with this API, but I thought it was interesting to see how you could store them and retrieve them in this model.

Reply:
And, I added a `CompoundAttribute` as well, to see how that would work. Incidentally, I think it would be pretty easy to add a `CompoundAttributeValue` to the existing APIs, so I don't think this particular feature is a value-add with the typed-key approach.

Reply:
@tylerbenson re-reading your comment, I realized I address half of what you were suggesting. Can you suggest how your proposed API would work, with the way that java generics function? I'm confused about how it would work, exactly.